[DependencyInjection] Allow environment variables with `.` in them

maxbaldanza opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 8.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? |no
| Issues        | Fix https://github.com/symfony/symfony/issues/33806#issuecomment-539480798 -< old discussion that says `.` could be added
| License       | MIT

In our system, environment variables are generated dynamically (e.g. by a Go controller) and use dot-separated names to represent hierarchical keys.

These variables are valid at the OS level and can be accessed via getenv()/$_ENV, but Symfony currently rejects them due to strict validation.

This change allows consuming such existing environment variables.

Commit:
[DependencyInjection] Allow environment variables with `.` in them

In our system, environment variables are generated dynamically (e.g., by a Go controller)
and use dot-separated names to represent hierarchical keys.

These variables are valid at the OS level and can be accessed via getenv()/$_ENV,
but Symfony currently rejects them due to strict validation.

This change allows consuming such existing environment variables.

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/CHANGELOG.md:
```diff
@@ -16,6 +16,7 @@ CHANGELOG
  * Deprecate XML configuration format, use YAML or PHP instead
  * Deprecate `ExtensionInterface::getXsdValidationBasePath()` and `getNamespace()`
  * Deprecate the fluent PHP format for semantic configuration, use `$container->extension()` or return an array instead
+ * Allow environment variables with `.` in them
```
should me move to a 8.1 section

maxbaldanza replied:
Thanks Nicolas.

Sorry for targeting the wrong branch. I've now updated 👍

nicolas-grekas reviewed (approved):
LGTM. This should target 8.1 since it's a new feature.

alexandre-daubois commented on src/Symfony/Component/DependencyInjection/Tests/ParameterBag/EnvPlaceholderParameterBagTest.php:
```diff
@@ -183,6 +183,16 @@ public function testGetEnvAllowsNull()
+        $bag->get('env(dynamic.var)');
+        $bag->resolve();
+
+        $this->assertEquals('foo', $bag->all()['env(dynamic.var)']);
```
```suggestion
        $this->assertSame('foo', $bag->all()['env(dynamic.var)']);
```

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/ParameterBag/EnvPlaceholderParameterBag.php:
```diff
@@ -41,8 +41,8 @@ public function get(string $name): array|bool|string|int|float|\UnitEnum|null
             }
-            if (!preg_match('/^(?:[-.\w\\\\]*+:)*+\w*+$/', $env)) {
-                throw new InvalidArgumentException(\sprintf('The given env var name "%s" contains invalid characters (allowed characters: letters, digits, hyphens, backslashes and colons).', $name));
+            if (!preg_match('/^(?:[-.\w\\\\]*+:)*+[\w.]*+$/', $env)) {
```
looking for `[-.\w` in the code base, I can find more lines that should be patched

maxbaldanza replied:
Do you mean for the `decryption_env_var` in `FrameworkExtension` and `PhpDumper`? Do you think these should also support a dot or were you looking elsewhere?

nicolas-grekas replied:
Of course, things should be consistent

maxbaldanza replied:
Hi @nicolas-grekas I've updated those two places. I couldn't see tests for these so added some, let me know if I missed anything.

Commit:
[FrameworkBundle] Add support for dots in `framework.secrets.decryption_env_var`

DependencyInjection now supports dots in environment variables so applying the same
change to the framework bundle for consistency
